### ISSUE TITLE
Reconstruction Scale/Pose Mismatch: Predicted Poses and Intrinsics Not Aligning with Ground Truth

### ISSUE THREAD

**Issue by LouisBdg** [facebookresearch/map-anything#93]
Hello, thank you for this great repository.
I’m tracking cameras and reconstructing a scene using this repository. I have ground truth poses and intrinsics for each image. When I run inference with only images, the point cloud looks reasonable but is incorrectly scaled. Adding intrinsics doesn’t change the result, but adding poses improves the scale—however, the point cloud becomes incoherent and unrecognizable. To find where this problem comes from, I looked at input/output difference in poses and intrinsics. 
I plotted the poses in 3D:
With images only: Point cloud is scaled incorrectly (4x too large). Predicted poses are in green, measured ones in blue.
 
<img width="400" height="359" alt="Image" src="https://github.com/user-attachments/assets/ef1e99b7-fb54-42e7-ae84-c717b7a7369f" />

With images plus intrinsics: Result is almost identical to image-only.

<img width="427" height="314" alt="Image" src="https://github.com/user-attachments/assets/0147e680-fab8-48d0-a43e-910bb50438ac" />
 
With images, intrinsics and poses: Scale improves, but is still incorrect, the point cloud is noisy and doesn’t match the scene, it seems like the pointclouds related to each input images duplicate the scanned object in different but nearby positions. My issue seems close to the #33 one.
 
<img width="393" height="340" alt="Image" src="https://github.com/user-attachments/assets/94e9557a-16d7-468f-a198-d80221e186e9" />

About intrinsics: 
The predicted intrinsics (fx, fy, cx, cy) are significantly different from the provided values. And this difference is not completely correlated to the reshape process of images (2048x1536) -> (518, 392) as seen in the issue #38 .
One of the differences is that input cx and cy are not centered while the predicted intrinsics are centered. I’ve seen a similar problem in issue #67 but with no solution for me.

Do somebody know where my errors come from? Is it normal to have predicted poses and intrinsics that differs that much from inputs? As somebody managed to have a satisfying result with poses and intrinsics as inputs?
Thank you for your help!

**Comment by shaashou** [comment]
Have you resolved these issues?

**Comment by Nik-V9** [comment]
Hi, thanks for testing MapAnything!

The quality of the pass through when using pose inputs depends on the testing domain since the model expects the pose scale to align with its internal metric scale prior (based on training data). For example, MapAnything has an issue with toy scale objects and aerial domain since they are not featured in the metric-scale training set.

When observing suboptimal performance using poses as input (the same case for COLMAP or up to scale poses as input), I would recommend testing out `ignore_pose_inputs=True` in the inference API. For example, you can see the effect of this flag here: https://github.com/facebookresearch/map-anything/issues/9#issuecomment-3312096039

**Comment by shaashou** [comment]
> Hi, thanks for testing MapAnything!
> 
> The quality of the pass through when using pose inputs depends on the testing domain since the model expects the pose scale to align with its internal metric scale prior (based on training data). For example, MapAnything has an issue with toy scale objects and aerial domain since they are not featured in the metric-scale training set.
> 
> When observing suboptimal performance using poses as input (the same case for COLMAP or up to scale poses as input), I would recommend testing out `ignore_pose_inputs=True` in the inference API. For example, you can see the effect of this flag here: [#9 (comment)](https://github.com/facebookresearch/map-anything/issues/9#issuecomment-33

> Hi, thanks for testing MapAnything!
> 
> The quality of the pass through when using pose inputs depends on the testing domain since the model expects the pose scale to align with its internal metric scale prior (based on training data). For example, MapAnything has an issue with toy scale objects and aerial domain since they are not featured in the metric-scale training set.
> 
> When observing suboptimal performance using poses as input (the same case for COLMAP or up to scale poses as input), I would recommend testing out `ignore_pose_inputs=True` in the inference API. For example, you can see the effect of this flag here: [#9 (comment)](https://github.com/facebookresearch/map-anything/issues/9#issuecomment-3312096039)

Hello, is it possible to obtain the same depth information as the actual value if I only input a single viewpoint? I've observed that the camera position parameter isn't important if it's a single viewpoint; I only plan to input RGB, camera intrinsics, and depth map.

**Comment by Nik-V9** [comment]
Hi @shaashou yes that should be possible.

I would recommend trying both True and False for the ignore_depth_scale_inputs flag.

**Comment by shaashou** [comment]
I will try your method, and if I encounter any problems again, I hope to receive your response

**Comment by roboyike** [comment]
My input image is 720x1280. How should I process it? No matter what I do now, I can't infer my correct intrinsic parameters.

> Hi [@shaashou](https://github.com/shaashou) yes that should be possible.你好，是的，这应该是可行的。
> 
> I would recommend trying both True and False for the ignore_depth_scale_inputs flag.我建议 ignore_depth_scale_inputs 旗时可以同时尝试“真”和“假”两个选项。

**Comment by Nik-V9** [comment]
Hi @roboyike are you talking about the intrinsics pass through? Note that the model requires 518 longest side images as input.

**Comment by shaashou** [comment]
> Hi [@roboyike](https://github.com/roboyike) are you talking about the intrinsics pass through? Note that the model requires 518 longest side images as input.

I am currently also using 720x1280 resolution RGB images with the actual camera's intrinsic parameters. Should I convert my images to 512x294x3 format before feeding them into the preprocess_input function?

**Comment by Nik-V9** [comment]
The model inference takes care of the resizing and the output will correspond to the resized input. You can check the shape of the predictions or viz RGB for resized shape. Please see: https://github.com/facebookresearch/map-anything/issues/38

**Comment by shaashou** [comment]
> The model inference takes care of the resizing and the output will correspond to the resized input. You can check the shape of the predictions or viz RGB for resized shape. Please see: [#38](https://github.com/facebookresearch/map-anything/issues/38)模型推理过程会自动处理图像大小调整，输出结果将与调整后的输入图像相对应。您可以查看预测结果的形状，或者查看调整后图像的 RGB 值。请参阅： [#38](https://github.com/facebookresearch/map-anything/issues/38)

I noticed this issue and found that it seems to eventually predict correctly, but my direct input of RGB and camera intrinsics, fx, fy, has significant differences, while cx and cy cannot be predicted correctly at all

**Comment by shaashou** [comment]
@Nik-V9

**Comment by shaashou** [comment]
> The model inference takes care of the resizing and the output will correspond to the resized input. You can check the shape of the predictions or viz RGB for resized shape. Please see: [#38](https://github.com/facebookresearch/map-anything/issues/38)


Hello, the first image shows the parameters of my data after being processed by the preprocess_inputs function, but after inference, the output does not match at all, just like the second image.

<img width="303" height="86" alt="Image" src="https://github.com/user-attachments/assets/3962d405-9947-43e1-8b7c-45f975b4ef8a" />


<img width="419" height="61" alt="Image" src="https://github.com/user-attachments/assets/e3d3b17f-9257-48e9-943a-99cacad8de46" />

**Comment by shaashou** [comment]
These are all my code and the images I tested

[demo_images_multi_inference.py](https://github.com/user-attachments/files/23596861/demo_images_multi_inference.py)

<img width="1280" height="720" alt="Image" src="https://github.com/user-attachments/assets/071587e1-47c8-4d70-9388-a3a4bf8799cf" />
Before the test, I had completed the eye-on-hand calibration and obtained the camera intrinsic matrix.
I very much hope to receive your reply

**Comment by shaashou** [comment]
> The model inference takes care of the resizing and the output will correspond to the resized input. You can check the shape of the predictions or viz RGB for resized shape. Please see: [#38](https://github.com/facebookresearch/map-anything/issues/38)模型推理过程会自动处理图像大小调整，输出结果将与调整后的输入图像相对应。您可以查看预测结果的形状，或者查看调整后图像的 RGB 值。请参阅： [#38](https://github.com/facebookresearch/map-anything/issues/38)

I solved this problem; it was caused by an incorrect setting of my camera's internal parameters. Thank you for your patient explanation. @Nik-V9

**Comment by joejep** [comment]
@shaashou Does the predicted pointcloud match the ground truth now?

**Comment by shaashou** [comment]
yes

> [@shaashou](https://github.com/shaashou) Does the predicted pointcloud match the ground truth now?

**Comment by joejep** [comment]
do you mind sharing the code and the dataset? . We can setup a zoom meeting

**Comment by shaashou** [comment]
> do you mind sharing the code and the dataset? . We can setup a zoom meeting

Sorry, I can't provide my dataset, but I forked this repository. I inferred it using depth maps and camera parameters, and it is accurate. My repository address: https://github.com/shaashou/map-anything-robotics

**Comment by shaashou** [comment]
> do you mind sharing the code and the dataset? . We can setup a zoom meeting你介意分享代码和数据集吗？我们可以安排一个 Zoom 会议

I tested the scale of reconstructing a desktop scene using a robotic arm eye-in-hand external camera.

**Comment by joejep** [comment]
Oh, I was thinking is a regular dataset. Would eth3d dataset be sufficient then. The multi input inference. I can see you passed images, intrinsics and depth. If depth isn’t available will the code work if I pass poses as well

**Comment by shaashou** [comment]
> Oh, I was thinking is a regular dataset. Would eth3d dataset be sufficient then. The multi input inference. I can see you passed images, intrinsics and depth. If depth isn’t available will the code work if I pass poses as well哦，我当时想的是常规数据集。那么 eth3d 数据集就足够了吗？关于多输入推理。我看到你传递了图像、内参和深度信息。如果没有深度信息，我再传递姿态信息，代码还能运行吗？

Running it is definitely possible, but it feels like with depth information, the scale would be more accurate.

**Comment by shaashou** [comment]
> Oh, I was thinking is a regular dataset. Would eth3d dataset be sufficient then. The multi input inference. I can see you passed images, intrinsics and depth. If depth isn’t available will the code work if I pass poses as well哦，我当时想的是常规数据集。那么 eth3d 数据集就足够了吗？关于多输入推理。我看到你传递了图像、内参和深度信息。如果没有深度信息，我再传递姿态信息，代码还能运行吗？

I haven't tried the ETH3D dataset; I only adapted it to the images I took myself.

**Comment by joejep** [comment]
> > Oh, I was thinking is a regular dataset. Would eth3d dataset be sufficient then. The multi input inference. I can see you passed images, intrinsics and depth. If depth isn’t available will the code work if I pass poses as well哦，我当时想的是常规数据集。那么 eth3d 数据集就足够了吗？关于多输入推理。我看到你传递了图像、内参和深度信息。如果没有深度信息，我再传递姿态信息，代码还能运行吗？
> 
> I haven't tried the ETH3D dataset; I only adapted it to the images I took myself.

Since you are getting depth information. Are you using realsense camera to get depth information real time. Or you you have an algorithm to get depth information from images. I wish if I can get the multi mode inference script including the poses would just solve my problem

**Comment by shaashou** [comment]
> > > Oh, I was thinking is a regular dataset. Would eth3d dataset be sufficient then. The multi input inference. I can see you passed images, intrinsics and depth. If depth isn’t available will the code work if I pass poses as well哦，我当时想的是常规数据集。那么 eth3d 数据集就足够了吗？关于多输入推理。我看到你传递了图像、内参和深度信息。如果没有深度信息，我再传递姿态信息，代码还能运行吗？哦，我以为是一个常规数据集。那么 eth3d 数据集就足够了。多输入推理。我可以看到你传递了图像、内在特征和深度。如果深度不可用，如果我通过姿势，代码也会工作哦，我当时想的是常规数据集。那么 eth3d 数据集就足够了吗？关于多输入推理。我看到你传递了图像、内参和深度信息。如果没有深度信息，我再传递姿势信息，代码还能运行吗？
> > 
> > 
> > I haven't tried the ETH3D dataset; I only adapted it to the images I took myself.我没有尝试过 ETH3D 数据集；我只是将其适配到我自己拍摄的图像上。
> 
> Since you are getting depth information. Are you using realsense camera to get depth information real time. Or you you have an algorithm to get depth information from images. I wish if I can get the multi mode inference script including the poses would just solve my problem既然您已经获取了深度信息，请问您是使用 RealSense 相机实时获取深度信息，还是使用某种算法从图像中提取深度信息？我希望如果能获得包含姿态信息的多模态推理脚本，就能解决我的问题。

Obtained using a RealSense camera

**Comment by shaashou** [comment]
> > > Oh, I was thinking is a regular dataset. Would eth3d dataset be sufficient then. The multi input inference. I can see you passed images, intrinsics and depth. If depth isn’t available will the code work if I pass poses as well哦，我当时想的是常规数据集。那么 eth3d 数据集就足够了吗？关于多输入推理。我看到你传递了图像、内参和深度信息。如果没有深度信息，我再传递姿态信息，代码还能运行吗？哦，我以为是一个常规数据集。那么 eth3d 数据集就足够了。多输入推理。我可以看到你传递了图像、内在特征和深度。如果深度不可用，如果我通过姿势，代码也会工作哦，我当时想的是常规数据集。那么 eth3d 数据集就足够了吗？关于多输入推理。我看到你传递了图像、内参和深度信息。如果没有深度信息，我再传递姿势信息，代码还能运行吗？
> > 
> > 
> > I haven't tried the ETH3D dataset; I only adapted it to the images I took myself.我没有尝试过 ETH3D 数据集；我只是将其适配到我自己拍摄的图像上。
> 
> Since you are getting depth information. Are you using realsense camera to get depth information real time. Or you you have an algorithm to get depth information from images. I wish if I can get the multi mode inference script including the poses would just solve my problem既然您已经获取了深度信息，请问您是使用 RealSense 相机实时获取深度信息，还是使用某种算法从图像中提取深度信息？我希望如果能获得包含姿态信息的多模态推理脚本，就能解决我的问题。

You can check out the demo_images_multi_inference.py file in my repository. If it helps you, I hope you can give it a star.

**Comment by joejep** [comment]
I will give it a star and if I use your repository. I will give you the citation and acknowledgement of your help in my paper as well. I want to ask about the intrinsic and poses. I know the transformer assume 518 or 512 or 514 by default which affect the intrinsic passed. What if my input image is say 1024 by 1024 which is not factor of 14. How do I keep it fix so that it doesn’t resize it and also it doesn’t affect the intrinsic passed

**Comment by shaashou** [comment]
> I will give it a star and if I use your repository. I will give you the citation and acknowledgement of your help in my paper as well. I want to ask about the intrinsic and poses. I know the transformer assume 518 or 512 or 514 by default which affect the intrinsic passed. What if my input image is say 1024 by 1024 which is not factor of 14. How do I keep it fix so that it doesn’t resize it and also it doesn’t affect the intrinsic passed我会给你点赞，如果我使用了你的代码库，我也会在我的论文中引用并感谢你的帮助。我想问一下关于内参和姿态的问题。我知道 Transformer 默认使用 518、512 或 514 像素，这会影响传递的内参。如果我的输入图像是 1024x1024 像素，不是 14 的倍数，我该如何保持图像大小不变，既不影响传递的内参，又能保证图像大小不受影响？

You can check the demo_colmap_multi_inference.py code in my repository. My initial resolution is also not a multiple of 14, but I kind of forgot.

**Comment by joejep** [comment]
You have referred me to two inference code now. Which works best

**Comment by shaashou** [comment]
> You have referred me to two inference code now. Which works best您现在给了我两个推理代码。哪个效果最好？

I am using demo_colmap_multi_inference.py. In this, I adapted my initial image resolution to 518×294 and exported it in COLMAP format.

**Comment by joejep** [comment]
What if the image size is large say 1024 by 1024 which is larger than the reference size of the transformer. Does it retain the actual intrinsic passed or resize it to that of the transformer

**Comment by shaashou** [comment]
> What if the image size is large say 1024 by 1024 which is larger than the reference size of the transformer. Does it retain the actual intrinsic passed or resize it to that of the transformer如果图像尺寸很大，比如 1024 x 1024，大于变压器的参考尺寸，它会保留实际传递的固有参数，还是会将其调整为变压器的尺寸？

Will adjust to the parameters corresponding to the resolution

**Comment by joejep** [comment]
Sorry for asking too much. What are the parameters that needs to be adjusted. I am thinking the inference code just takes images , poses and intrinsic and sometimes depth as input.

**Comment by joejep** [comment]
But the network predicts the intrinsic except you have to manually calculate it and replace it with the one the network computes. Is that what should be done. What about the fixed size of the image that the transformser accepts

**Comment by shaashou** [comment]
> Sorry for asking too much. What are the parameters that needs to be adjusted. I am thinking the inference code just takes images , poses and intrinsic and sometimes depth as input.不好意思问得太多。需要调整哪些参数？我猜推理代码的输入应该是图像、姿态、内参，有时还有深度信息。

You can take a look at the preprocess_inputs function in the code; I think it adjusts the inputs to the model's size.

<img width="971" height="667" alt="Image" src="https://github.com/user-attachments/assets/2fc80bc0-0e05-4cba-bd02-e893489947e1" />

**Comment by joejep** [comment]
We don’t want the iinput image and intrinsic resized. If it would be resized we want the predicted depth to match the ground truth depth which would be same size as 1024 by 1024.

**Comment by shaashou** [comment]
> We don’t want the iinput image and intrinsic resized. If it would be resized we want the predicted depth to match the ground truth depth which would be same size as 1024 by 1024.我们不希望输入图像和内参被调整大小。如果必须调整大小，我们希望预测的深度与真实深度相匹配，即尺寸为 1024x1024。

Actually, what I want to do is the same as what you're trying to do. My code realizes my ideas, and I've experimented with the robotic arm at real scale. You can take a look at my code; I hope it can be helpful to you.

**Comment by joejep** [comment]
Thank you so much
